File: cpp/hardware/storage/SSD.cpp
#pragma once
#include <bits/stdc++.h>

#include "StorageDevice.cpp"

class SSD : public StorageDevice
{
private:
    std::string nandType;
    int tbw; // Total Bytes Written

public:
    // Constructor
    SSD() : StorageDevice(), nandType(""), tbw(0) {}
    SSD(const std::string &serial, const Date &installDate, bool installed, int capacityGB, int readSpeedMBs, int writeSpeedMBs, const std::string &id, const std::string &manufacturer, const std::string &model, double price, std::string nandType, int tbw)
        : StorageDevice(serial, installDate, installed, capacityGB, readSpeedMBs, writeSpeedMBs, id, manufacturer, model, price), nandType(nandType), tbw(tbw) {}

    // Destructor
    ~SSD() {}

    // Setters
    void setNandType(const std::string &nandType)
    {
        this->nandType = nandType;
    }

    void setTbw(int tbw)
    {
        this->tbw = tbw;
    }

    // Getters
    std::string getNandType() const
    {
        return nandType;
    }

    int getTbw() const
    {
        return tbw;
    }

    bool connect() override
    {
        printf("SSD connected\n");
        return true;
    }

    bool disconnect() override
    {
        printf("SSD disconnected\n");
        return true;
    }
};

File: python/hardware/storage/ssd.py
from datetime import date
from hardware.storage.storage_device import StorageDevice

class SSD(StorageDevice):
    def __init__(self, serial_number="", installation_date=None, physically_installed=False, capacity_gb=0, read_speed_mbs=0, write_speed_mbs=0, id="", manufacturer="", model="", price=0.0, nand_type="", tbw=0):
        super().__init__(serial_number, installation_date, physically_installed, capacity_gb, read_speed_mbs, write_speed_mbs, id, manufacturer, model, price)
        self._nand_type = nand_type
        self._tbw = tbw

    # Setters
    def set_nand_type(self, nand_type):
        self._nand_type = nand_type

    def set_tbw(self, tbw):
        self._tbw = tbw

    # Getters
    def get_nand_type(self):
        return self._nand_type

    def get_tbw(self):
        return self._tbw

    def connect(self):
        print("SSD connected")
        return True

    def disconnect(self):
        print("SSD disconnected")
        return True